### ISSUE TITLE
In Firefox, NVDA causes the first link in the DOM to be skipped intermittently (the second link/focusable element is focused)

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#3575]
**Reported by schne324 on 2013-10-11 18:44**
take the example code here:

```
<!DOCTYPE html>
<head>
    <title>NVDA Proof of Bug</title>
    <style type="text/css">
    :focus {
        outline: 2px solid #f00;
    }
    </style>
</head>
<body>
    <a href="#">First Link</a>
    <a href="#asdf">SECOND Link</a>
    <div id="asdf">asdf</div>
</body>
</html>
```

I am running the most recent update of Firefox and I am seeing the first link not receiving focus every time.  This is a problem for pages with skip links as the first child of the body.

**Comment by hidanielle** [comment]
I'm experiencing this issue currently and have tested without sayAll.

**Comment by hidanielle** [comment]
I've tested this with SayAll enabled and disabled. With SayAll disbaled, NVDA will read out that you are focused on the first skip link and it is functional, but it does not receive visual focus. I believe this is still a bug, as without NVDA the buttons both receive visual focus. The button should receive visual focus in all cases.

**Comment by Adriani90** [comment]
@schne324, @hidanielle is this still an issue in last NVDA Alpha?
cc: @leonardder

**Comment by Adriani90** [comment]
@Qchristensen we need your helpful eyes again. :-) Could you please test with NVDA last alpha or 2019.1.1 in different browsers if the first link gets visual focus, as it is claimed in this issue? Thanks.

**Comment by Qchristensen** [comment]
In both Firefox and Chrome, what I found was:
- If I refresh the page with F5, while the focus is on the second link, the page refreshes and the focus stays on the second link.  Pressing TAB moves into the browser tab / menu area (as it usually does once you Tab past the last control on a page).
- If NVDA is not running, when the page refreshes, neither link has focus and TAB takes focus to the first link.
- If the focus is in the address bar and ENTER is pressed to load the page (as long as you haven't changed anything), the page loads, the focus stays in the address bar, and TAB moves to the first link.

One other thing I did find however, In Firefox, if the focus is in the address bar and you press F5 to refresh the current page, when you then press TAB, the focus moves through the firefox menu etc controls, and then jumps straight to the second link.  I couldn't replicate that with Chrome, or with NVDA not running.

**Comment by erickbelfy** [comment]
With a given HTML code:

```
<nav>
<ul class="css-bnahpf">
   <li class="css-10tjp12">
       <button class="MuiButtonBase-root MuiButton-root css-1hk3q50 MuiButton-text" 
                      tabindex="0" 
                      type="button" 
                      aria-pressed="true"
          >
             <span class="MuiButton-label">
                     <span>Disable accessibility mode</span>
              </span>
              <span class="MuiTouchRipple-root"></span>
         </button>
    </li>
   <li class="css-10tjp12">
           <button 
               class="MuiButtonBase-root MuiButton-root css-1hk3q50 MuiButton-text" 
               tabindex="0" 
               type="button"
            >
                 <span class="MuiButton-label">
                         <span>Skip to document</span>
               </span>
               <span class="MuiTouchRipple-root"></span>
            </button>
      </li>
</ul>
</nav>
<header>
     <div>
         <div>
               <button class="hamburger-button" aria-label="main menu"> 
                     <span class="MuiButton-label"><svg ></svg></span>
                     <span class="MuiTouchRipple-root"></span>
              </button>
         </div>
    </div>
</header>
```

With NVDA on, when i hit tab instead of going to the first nav->ul->li->button it goes to the header first button (hamburger-button), even with css turned off i can't reach the first button while NVDA on.

**Comment by Adriani90** [comment]
@jcsteh is this not the same behavior like in #4386? If yes, then I suggest to close #4386 in favor of this one because here we have more recent updates and code examples. What do you think?

**Comment by masi** [comment]
It's the setting "Enable browse mode on page load" that triggers the behaviour. In browse mode NVDA reads out the page title and the "first line" (?), which happens to be the link. A tab will focus the link and bring it into via via CSS, but NVDA does not repeat the text.
When I disable browse mode on page load NVDA will read out the page title and stop.

**Comment by fatimaazizz** [comment]
Have same  issue as mentioned above css focus  is visible on tab but NVDA remains silent .than tab to next 2nd element.
came back to element  with shift + tab than NVDA reads it. this working fine chromevox.

**Comment by Adriani90** [comment]
@michaelDCurran does  PR #14611 maybe also affect this positively?

**Comment by Adriani90** [comment]
I have slightly modified the syntax of the initial code snippet in the isue description as follows:
```
data:text/html,<a href="">First Link</a><a href="asdf">SECOND Link</a><div id="asdf"asdf</div>
```

Using this example, the issue is not reproducible with NVDA 2025.1.2 RC in Firefox 139 or Chromium 139.
I am closing as works for me.
If you are still having this issue, please comment with an updated minimal test case and we can reopen.

**Comment by mrazzari** [comment]
Tested NVDA 2025.1.2 RC1, this is still an issue. 

Slightly revised test case:
```html
data:text/html,<a href="asdf">First Link</a> <a href="asdf">SECOND Link</a><p>testing</p>
```
Same as above from @Adriani90 , only added `asdf` to first link for consistency, and replaced the broken div for a paragraph.

Repro:

1. Copy snippet to clipboard
3. Ctrl+T to Open new tab
4. Ctrl+V to paste snippet
5. Enter to 'navigate' to it
6. Hit TAB only once < focus is now at the second link 👎 

Tested Firefox 139 and Chrome 137.

Now here's the thing: 
* if I allow NVDA to finish saying the entire page before TAB (Step 6), it will focus on the **first** link
* if I hit TAB quickly after Enter (step 5 to 6), that is, while NVDA is still reading, it will focus on the **second** link
* If I remove the test paragraph (leaving just the 2 links), it will always focus on the **second** link regardless of how quickly I tab into the content.

**Comment by Adriani90** [comment]
@mrazzari you said:

> • if I allow NVDA to finish saying the entire page before TAB (Step 6), it will focus on the first link.

It seems it has been implemented like that in NVDA for say all.
I think it is easier to understand with following test case:
```
data:text/html,<a href="asdf">First Link</a> <a href="asdf">SECOND Link</a><a href="asdf">third Link</a><a href="asdf">fourth Link</a>
```

1. When you press f5 and NVDA starts reading the page, pressing nvda+tab reports only "document protected" because there is no system focus set yet. NVDA is acting in its virtual document.
2. When say all finishes reading and you press tab, the first system focus event is set on the first link, and the virtual cursor will land at this position as well which is the top of the virtual document. It is wrapping to the top.
3. When say all starts and you press tab very quickly, the system focus will catch up with the virtual cursor at the position where say all is interupted (e.g. second link, third link, etc.).
4. When say all finishes reading and you press shift+tab, you land on the position before the last link, this is third link.
* If you set a <p> after fourth link, shift+tab at the end of say all will probably set the system focus to the next focussable element which would be fourth link.
* If say all ends on a focusable element like a link, then pressing shift+tab will set the system focus at the position of the virtual cursor and this is the penultimate focussable element.
5. If you press f5 and tab very quickly, system focus will be set on the second focussable element as long as say all is still reading first link.
6. If you press f5 and press shift+tab very quickly, the virtual cursor will wrap to the bottom of the virtual document and will land on fourth link. System focus will be set to fourth link.

So in short, say all can read anything on the page, even if it is not focussable with system focus. Pressing tab or shift+tab will interupt say all and will trigger a focus event for virtual cursor and system focus,. They will sync with each other to the nearest next / previous focussable element possible, or will wrap to the top or the bottom of the virtual document depending when you press tab or shift+tab.

> • If I remove the test paragraph (leaving just the 2 links), it will always focus on the second link regardless of how quickly I tab into the content.

I cannot reproduce that. Removing the paragrpaph leads to always focusing first link if you let say all finish reading before pressing tab or shift+tab. This happens because virtual cursor wraps to the top of the virtual document when it finishes reading and you press tab and it happens that first link in your case is also the penultimate focussable element after say all finishes reading so you land on first link as well after pressing shift+tab.
Doing this very quickly will always bring you to second link.

**Comment by Adriani90** [comment]
So I think your expected behavior is as follows:
1. not wrap the virtual cursor to top or bottom of the virtual document but leave it completely when pressing tab after say all finishes reading and
2. not wrap the virtual document but leave it completely when you press shift+tab while say all is still reading the first element on the page.

cc: @jcsteh could you give some thoughts on why virtual cursor wraps to the top or bottom of the virtual document in this case? Was this intentionally implemented back then?

**Comment by jcsteh** [comment]
If you're at the top of the document and the first link is right at the start of the document, you're already on the first link,so tab moves to the second. If you're at the bottom of the document (which will happen after say all), there is nothing focusable after the cursor, so tab is passed through to the browser directly.
